File: 剑指offer/C++_Version/03. 数组中重复的数字.cpp
//方法1
class Solution {
public:
    int findRepeatNumber(vector<int>& nums) {
        int i = 0;
        while(i < nums.size()){
            if(nums[i]==i){
                i++;
                continue;
            }
            if(nums[nums[i]]==nums[i]){
                return nums[i];
            }
            swap(nums[i], nums[nums[i]]);
        }
        return -1;
    }
};


//方法2
class Solution {
public:
    int findRepeatNumber(vector<int>& nums) {
        unordered_map<int, bool> map;
        for(int num: nums){
            if(map[num]) return num;
            map[num] = true;
        }
        return -1;
    }
};

File: 剑指offer/Java_Version/03. 数组中重复的数字.java
//方法1
class Solution {
    public int findRepeatNumber(int[] nums) 
    {
        Set<Integer> dic = new HashSet<>();
        for(int num : nums)
        {
            if(dic.contains(num))
                return num;
            dic.add(num);
        }
        return -1;
    }
}



//方法2
class Solution {
    public int findRepeatNumber(int[] nums) 
    {
        int i = 0;
        while(i<nums.length)
        {
            if(nums[i]==i)
            {
                i+=1;
                continue;
            }
            if(nums[nums[i]] == nums[i])
                return nums[i];
            int tmp = nums[i];
            nums[i] = nums[tmp];
            nums[tmp] = tmp; 
        }
        return -1;
    }
}
